nixos/modules/services/misc/matrix-appservice-irc.nix

```diff
@@ -0,0 +1,123 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  dataDir = "/var/lib/matrix-appservice-irc";
+in {
+  options = {
+    services.matrix-appservice-irc = {
+      enable = mkEnableOption "a Node.js IRC bridge for Matrix";
+
+      port = mkOption {
+        type = types.port;
+        default = 9999; # from https://github.com/matrix-org/matrix-appservice-irc/blob/0.12.0/README.md#4-running
+        description = ''
+          Port number on which the bridge should listen for internal communication with the Matrix homeserver.
+        '';
+      };
+
+      config = mkOption rec {
+        type = types.attrs;
+        apply = recursiveUpdate default;
+        default = {
+          ircService = {
+            databaseUri = "nedb://${dataDir}/data";
+            passwordEncryptionKeyPath = "${dataDir}/passkey.pem";
+          };
+        };
+        example = literalExample ''
+          {
+            homeserver = {
+              url = "http://localhost:8008";
+              domain = "localhost";
+            };
+            ircService.servers = {
+              "irc.example.com" = {
+                name = "ExampleNet";
+                dynamicChannels.groupId = "+myircnetwork:localhost";
+              };
+            };
+          }
+        '';
+        description = ''
+          <filename>config.yaml</filename> configuration as a Nix attribute set.
+
+          Configuration options should match those described in
+          <link xlink:href="https://github.com/matrix-org/matrix-appservice-irc/blob/0.12.0/config.sample.yaml">
+          config.sample.yaml</link>.
+
+          Secret tokens should be specified using <option>environmentFile</option>
+          instead of this world-readable attribute set.
+        '';
+      };
+
+      localpart = mkOption {
+        type = types.string;
+        default = "ircbot";
+        description = ''
+          The user ID localpart to assign to the application service bot.
+        '';
+      };
+
+      homeserverService = mkOption {
+        type = types.nullOr types.string;
+        default = "matrix-synapse.service";
+        description = ''
+          Matrix homeserver service to wait for when starting the application service.
+        '';
+      };
+    };
+  };
+
+  config = let
+    cfg = config.services.matrix-appservice-irc;
+    configFile = pkgs.writeText "matrix-appservice-irc-config.json" (builtins.toJSON cfg.config);
+    registrationFile = pkgs.runCommand "matrix-appservice-irc-registration.yaml" { preferLocalBuild = true; } ''
+      ${pkgs.matrix-appservice-irc}/bin/matrix-appservice-irc \
```

Comment:
`pkgs` for modules is a package set for the target machine. So when you're cross-building on a machine with a different cpu arch this will probably break. See e.g. #62853 (tl;dr `pkgs.buildPackages` does the trick).

Reply:
Amended.